[rb] Upgrade to Ruby 3.2

p0deje opened the pull request:
### **User description**

1. Ruby 3.1 is EOL, so I'm bumping the minimum required version to 3.2. 
2. This forces us to switch to JRuby 10, because JRuby 9 is 3.1 compliant. 
3. JRuby 10 requires Java 21, so I had to upgrade Java runtime we use in Bazel. 
4. It caused some issues with Spotbugs, so I had to improve it (thanks @shs96c - https://seleniumhq.slack.com/archives/CBH302726/p1746696840466079?thread_ts=1746657055.463059&cid=CBH302726).
5. Spotbugs complained about some issues in Java tree. I have no idea what I was doing but it's now green. It would be great if Java folks could review.


___

### **PR Type**
Enhancement


___

### **Description**
- Upgrade minimum supported Ruby version to 3.2
  - Update `.ruby-version`, gemspecs, and documentation
  - Update RuboCop config for Ruby 3.2

- Update CI workflows to use latest Ruby, JRuby, TruffleRuby
  - Bump JRuby to 10.0.0.0 and TruffleRuby to 24.2.1
  - Update Ruby matrix in CI to 3.2.8 and 3.4.3

- Update Bazel and Ruby dependencies
  - Bump `rules_ruby` and various Ruby gem versions/checksums


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>selenium-devtools.gemspec</strong><dd><code>Require Ruby >= 3.2 in gemspec</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15714/files#diff-8ad4ad94aa5d5a9925dd12176bdb2486cea3a48c014faddf8673c2969985b740">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>selenium-webdriver.gemspec</strong><dd><code>Require Ruby >= 3.2 in gemspec</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15714/files#diff-f7c3979ae889cfb6e1b2386569a8bcd4f45e9a7e03d8c13bad52eb502a240aad">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td><strong>ci-rbe.yml</strong><dd><code>Update JRuby version in CI workflow to 10.0.0.0</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15714/files#diff-9ad6de1290c9f9260ba7c459ff1aa974c7c15aeeaae6f621aeda5dcde357b30e">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>ci-ruby.yml</strong><dd><code>Update Ruby, JRuby, TruffleRuby versions in CI matrix</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15714/files#diff-e71b669ae33da2f4e6f9e0858c13b326eaea0388934d0a21e3ebc31005b7aa90">+6/-6</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>pre-release.yml</strong><dd><code>Use Ruby 3.2 in pre-release workflow</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15714/files#diff-8f0c4d810b263ab478547020c847ee0297cc98c11c67fc2209ce604e52719607">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>.rubocop.yml</strong><dd><code>Set RuboCop TargetRubyVersion to 3.2</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15714/files#diff-010ef45e983d9d8b2649f7c517543a64e66362de0d8ef845df7527c634496e0f">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>.ruby-version</strong><dd><code>Set Ruby version to 3.2.8</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15714/files#diff-de119050d5da156e6f7cbb4c2fcf2a3c3366f9247254b03650a916bbfd2b574f">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Dependencies</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>MODULE.bazel</strong><dd><code>Update rules_ruby and Ruby gem dependencies</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15714/files#diff-6136fc12446089c3db7360e923203dd114b6a1466252e71667c6791c20fe6bdc">+37/-40</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>README.md</strong><dd><code>Update documentation for Ruby >= 3.2 support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15714/files#diff-1c388e87892816c537a0d0022226f4d7a44df02f45866f0dd0506f20fd0fddca">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

p0deje committed:
[rb] Upgrade to Ruby 3.2

- CRuby 3.1 is EOL
- JRuby 10 is the latest
- TruffleRuby 24.2 is the latest

p0deje committed:
[java] Fix spotbugs issues

titusfortner commented:
Hmm @cgoldberg any idea why the python remote tests don't like Java 21?

cgoldberg commented:
I have no idea. What's failing? I use Java 17 on my local, but I can try 21.

Edit: ok.. I see the errors. I'll look into it this afternoon.

cgoldberg commented:
I ran the remote tests in trunk with Java 21 (not using bazel, not using this PR branch), and everything ran fine.

In the CI logs for this PR, there are a ton of errors that occur when calling `driver.quit()`:

```
java.lang.ClassCastException: class jdk.internal.net.http.HttpClientFacade cannot be cast to class java.lang.AutoCloseable (jdk.internal.net.http.HttpClientFacade is in module java.net.http of loader 'platform'; java.lang.AutoCloseable is in module java.base of loader 'bootstrap')
	at org.openqa.selenium.remote.http.jdk.JdkHttpClient.close(JdkHttpClient.java:526)
	at org.openqa.selenium.remote.tracing.TracedHttpClient.close(TracedHttpClient.java:62)
	at org.openqa.selenium.grid.node.config.DriverServiceSessionFactory$1.stop(DriverServiceSessionFactory.java:209)
	at org.openqa.selenium.grid.node.DefaultActiveSession.execute(DefaultActiveSession.java:64)
	at org.openqa.selenium.grid.node.local.SessionSlot.execute(SessionSlot.java:130)
	at org.openqa.selenium.grid.node.local.LocalNode.executeWebDriverCommand(LocalNode.java:718)
	at org.openqa.selenium.grid.node.ForwardWebDriverCommand.execute(ForwardWebDriverCommand.java:48)
	at org.openqa.selenium.remote.http.Route$PredicatedRoute.handle(Route.java:397)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
	at org.openqa.selenium.remote.tracing.SpanWrappedHttpHandler.execute(SpanWrappedHttpHandler.java:87)
	at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
	at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
	at org.openqa.selenium.grid.node.Node.execute(Node.java:296)
	at org.openqa.selenium.grid.web.CombinedHandler.execute(CombinedHandler.java:55)
	at org.openqa.selenium.grid.web.RoutableHttpClientFactory$1.execute(RoutableHttpClientFactory.java:71)
	at org.openqa.selenium.grid.web.ReverseProxyHandler.execute(ReverseProxyHandler.java:91)
	at org.openqa.selenium.grid.router.HandleSession.execute(HandleSession.java:183)
	at org.openqa.selenium.remote.http.Route$PredicatedRoute.handle(Route.java:397)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
	at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
	at org.openqa.selenium.grid.router.Router.execute(Router.java:89)
	at org.openqa.selenium.grid.web.CheckOriginHeader.lambda$apply$0(CheckOriginHeader.java:69)
	at org.openqa.selenium.grid.web.CheckContentTypeHeader.lambda$apply$0(CheckContentTypeHeader.java:75)
	at org.openqa.selenium.grid.web.EnsureSpecCompliantResponseHeaders.lambda$apply$0(EnsureSpecCompliantResponseHeaders.java:34)
	at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
	at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
	at org.openqa.selenium.remote.AddWebDriverSpecHeaders.lambda$apply$0(AddWebDriverSpecHeaders.java:35)
	at org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)
	at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
	at org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)
	at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
	at org.openqa.selenium.netty.server.SeleniumHandler.lambda$channelRead0$0(SeleniumHandler.java:49)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

I don't know what they mean.

cgoldberg commented:
I would guess this change is causing it:

https://github.com/SeleniumHQ/selenium/pull/15714/files#diff-8c4bb26a4a1da292af9347bac545d0a98344057ae0373642be830b5744f828aaR526

p0deje commented:
@cgoldberg I think it's caused by my changes, let me revert and see if it helps. As I said, I have no idea what I do in Java code.

p0deje committed:
[java] fix autoclose casting

java.lang.ClassCastException: class jdk.internal.net.http.HttpClientFacade cannot be cast to class java.lang.AutoCloseable (jdk.internal.net.http.HttpClientFacade is in module java.net.http of loader 'platform'; java.lang.AutoCloseable is in module java.base of loader 'bootstrap'

p0deje commented:
@cgoldberg @titusfortner All green now.
